### PR TITLE
refactor: keep source location information in compiler state

### PR DESCRIPTION
**Pull request by saper**

### CODE REVIEW

**Comment by saper** [comment]
There are few things I am not sure of, like why `tokenize()` was sometimes called with a line number zero or how to handle the difference between two variants of `add_code()`

File: src/ldpl.cpp [hunk 1]
```diff
@@ -316,6 +316,10 @@ void warning(const string & msg)
 }
 
 //Shows error message and exits
+void badcode(const string & msg, const code_location where)
+{
+    error("unknown escape sequence (\033[0m" + where.current_file + ":" + to_string(where.current_line) + "\033[1;31m)");
```
Comment (dgarroDC) [inline src/ldpl.cpp:321]: `"unknown escape sequence"` -> `msg`

File: src/ldpl.cpp [hunk 2]
```diff
@@ -1091,371 +1094,372 @@ void compile_line(vector<string> & tokens, unsigned int line_num, compiler_state
...
+            badcode("CREATE STATEMENT statement inside IF", state.where);
         else if(state.closing_loop())
-            error("CREATE STATEMENT statement inside WHILE or FOR (\033[0m" + current_file + ":"+ to_string(line_num)+"\033[1;31m)");
+            badcode("CREATE STATEMENT statement inside WHILE or FOR", state.where);
         string model_line = tokens[2].substr(1, tokens[2].size() - 2);
         vector<string> model_tokens;
         vector<string> parameters = state.subprocedures[tokens[4]];
         trim(model_line);
-        tokenize(model_line, 0, model_tokens, state.current_file, true, ' ');
+        state.where.current_line = 0; // XXX why ?
```
Comment (dgarroDC) [inline src/ldpl.cpp:1461]: You can delete this line, 0 as the line number was wrong.

File: src/ldpl.h [hunk 3]
```diff
@@ -42,11 +46,15 @@ struct compiler_state{
     void add_var_code(string code){
         this->variable_code.push_back(code);
     }
-    void add_code(string code, unsigned int line_num = 0){
-	    auto& output = current_subprocedure == "" ? this->output_code : this->subroutine_code;
+    void add_code(string code) {
+        auto& output = current_subprocedure == "" ? this->output_code : this->subroutine_code;
+        output.push_back(code);
+    }
+    void add_code(string code, code_location where) {
```
Comment (dgarroDC) [inline src/ldpl.h:53]: I think you can delete `where` and use `this->where`.

File: src/ldpl.cpp (hunk 2 above)
Comment (saper) [reply src/ldpl.cpp]: done

File: src/ldpl.h (hunk 3 above)
Comment (saper) [reply src/ldpl.h:53]: I'll leave it for now. I was thinking about proper use of `#line` and there are some cases of calling `add_code(code)` without line number where we probably do not want to mess with `#line` once this feature is enabled. For now those two variants of `add_code` do the same thing but I hope we can enable it soon.

**Comment by Lartu** [comment]
Wonderful refactor! Thank you very, very much! Sorry for taking so long, I've been super busy.

**Pull request merged**